src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_ARRAYLIST_SIZE = 10;
+    private int size;
+
+    private T[] array;
+
+    public ArrayList() {
+        array = (T[])(new Object[DEFAULT_ARRAYLIST_SIZE]);
+    }
+
     @Override
     public void add(T value) {
-
+        if (size + 1 > array.length) {
+            growArray();
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        outOfRangeCheck((index == size && size != 0) ? (index - 1) : index);
+        if (size + 1 > array.length) {
+            growArray();
+        }
+        System.arraycopy(array, index, array, index + 1,array.length - index - 1);
+        array[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        while (size + list.size() > array.length) {
+            growArray();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        outOfRangeCheck(index == size ? index + 1 : index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        outOfRangeCheck(index == size ? index + 1 : index);
+        array[index] = value;
+        if (index > size) {
+            size = index;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        outOfRangeCheck(index == size ? index + 1 : index);
+        T result = array[index];
+        System.arraycopy(array,index + 1, array, index,array.length - index - 1);
+        size--;
+        return result;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        boolean isFound = false;
+        for (int i = 0; i < size; i++) {
+            if (array[i] == element || element != null && element.equals(array[i])) {
+                remove(i);
+                isFound = true;
+                break;
+            }
+        }
+        if (!isFound) {
+            throw new NoSuchElementException();
+        }
+        return element;
```

Comment:
```suggestion
        for (int i = 0; i < size; i++) {
            if (array[i] == element || element != null && element.equals(array[i])) {
                return remove(i);
            }
        }
        throw new NoSuchElementException("Element " + element + " is not present");
```
when the element is found in the array list - remove it, and return
then just throw an unchecked exception in the end, compiler understands that if there would occur an unchecked exception return statement is not needed because the program would not proceed to the next line
make exceptions informative